### PR TITLE
*: support global memory control for tidb

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/37816

Problem Summary:

### What is changed and how it works?

1. Add system variable `tidb_server_memory_limit` and `tidb_server_memory_limit_sess_min_size`
2. When tidb memory usage is larger than limit, kill the memory usage top1 sql.
3. After the top1 sql is killed, try to GC immediately.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support global memory control
```

### CODE REVIEW

File: util/expensivequery/expensivequery.go [hunk 1]
```diff
@@ -57,6 +58,7 @@ func (eqh *Handle) Run() {
 	defer ticker.Stop()
 	sm := eqh.sm.Load().(util.SessionManager)
 	record := &memoryUsageAlarm{}
+	serverMemoryQuota := &serverMemoryQuota{}
```
Comment (XuHuaiyu) [inline util/expensivequery/expensivequery.go:61]: Do we need to use an individual goroutine to monitor the global memory status? The expensive_query worker handles too many things

File: util/servermemoryquota/servermemoryquota.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		}
+	}
+}
+
+type serverMemoryQuota struct {
+	sqlStartTime time.Time
+	sessionID    uint64
+}
+
+func (s *serverMemoryQuota) CheckQuotaAndKill(bt uint64, sm util.SessionManager) {
```
Comment (XuHuaiyu) [inline util/servermemoryquota/servermemoryquota.go:66]:
```suggestion
killSessIfNeeded(s *sessionToBeKilled, bt uint64, sm util.SessionManager) *sessionToBeKill 
```

File: util/servermemoryquota/servermemoryquota.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		case <-smqh.exitCh:
+			return
+		}
+	}
+}
+
+type serverMemoryQuota struct {
+	sqlStartTime time.Time
+	sessionID    uint64
+}
```
Comment (XuHuaiyu) [inline util/servermemoryquota/servermemoryquota.go:64]:
```suggestion
type sessionToBeKilled struct {
        isKilling bool
	sqlStartTime time.Time
	sessionID    uint64
}
```

File: util/memory/tracker.go [hunk 4]
```diff
@@ -354,8 +366,11 @@ func (t *Tracker) Consume(bs int64) {
 	if bs == 0 {
 		return
 	}
-	var rootExceed, rootExceedForSoftLimit *Tracker
+	var rootExceed, rootExceedForSoftLimit, sessionTracker *Tracker
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:369]: This variable is somewhat strange, I think we can remove this var, and check `tracker.IsSession` below.

File: util/memory/tracker.go [hunk 5]
```diff
@@ -390,12 +405,32 @@ func (t *Tracker) Consume(bs int64) {
...
-	if bs > 0 && rootExceedForSoftLimit != nil {
-		tryAction(&rootExceedForSoftLimit.actionMuForSoftLimit, rootExceedForSoftLimit)
+	if bs > 0 && sessionTracker != nil {
+		// Kill the Top1 session
+		if sessionTracker.IsKilled.Load() {
+			tryAction(&sessionTracker.actionMuForHardLimit, sessionTracker)
+		}
+		// Update the Top1 session
+		memUsage := sessionTracker.BytesConsumed()
+		if limitSessMinSize := ServerMemoryLimitSessMinSize.Load(); limitSessMinSize > 0 {
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:415]: Why check `limitSessMinSize > 0`?

File: util/memory/tracker.go [hunk 6]
```diff
@@ -390,12 +405,32 @@ func (t *Tracker) Consume(bs int64) {
 		}
 	}
 
-	if bs > 0 && rootExceedForSoftLimit != nil {
-		tryAction(&rootExceedForSoftLimit.actionMuForSoftLimit, rootExceedForSoftLimit)
+	if bs > 0 && sessionTracker != nil {
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:408]: Put this if-branch to line 369.

File: util/servermemorylimit/servermemoryquota.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	if bt == 0 {
+		return
+	}
+	instanceStats := &runtime.MemStats{}
+	runtime.ReadMemStats(instanceStats)
+	if instanceStats.HeapInuse > bt {
+		t := memory.MemUsageTop1Tracker.Load()
+		if t != nil && uint64(t.BytesConsumed()) > memory.ServerMemoryLimitSessMinSize.Load() {
```
Comment (XuHuaiyu) [inline util/servermemorylimit/servermemoryquota.go:86]: Why check `uint64(t.BytesConsumed()) > memory.ServerMemoryLimitSessMinSize.Load()`

File: util/memory/tracker.go [hunk 8]
```diff
@@ -71,11 +77,14 @@ type Tracker struct {
...
-	bytesConsumed int64            // Consumed bytes.
-	bytesReleased int64            // Released bytes.
-	maxConsumed   atomicutil.Int64 // max number of bytes consumed during execution.
-	isGlobal      bool             // isGlobal indicates whether this tracker is global tracker
+	label               int              // Label of this "Tracker".
+	bytesConsumed       int64            // Consumed bytes.
+	bytesReleased       int64            // Released bytes.
+	maxConsumed         atomicutil.Int64 // max number of bytes consumed during execution.
+	SessionID           uint64           // SessionID indicates the sessionID the tracker is bound.
+	IsKilled            atomic.Bool      // IsKilled indicates whether this session is killed because OOM
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:85]: IsKilled --> needKill

File: util/servermemorylimit/servermemoryquota.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
+// Copyright 2022 PingCAP, Inc.
```
Comment (XuHuaiyu) [inline util/servermemorylimit/servermemoryquota.go:1]: rename this go file

File: util/servermemorylimit/servermemoryquota.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+// SetSessionManager sets the SessionManager which is used to fetching the info
+// of all active sessions.
+func (smqh *Handle) SetSessionManager(sm util.SessionManager) *Handle {
+	smqh.sm.Store(sm)
+	return smqh
+}
+
+// Run starts a server memory limit checker goroutine at the start time of the server.
+func (smqh *Handle) Run() {
```
Comment (XuHuaiyu) [inline util/servermemorylimit/servermemoryquota.go:45]: Add some comments for how this function works.

**Review by XuHuaiyu — APPROVED** [review]

File: executor/executor_test.go [hunk 11]
```diff
@@ -6063,3 +6063,66 @@ func TestTableLockPrivilege(t *testing.T) {
...
+	}
+	dom.ServerMemoryLimitHandle().SetSessionManager(sm)
+	go dom.ServerMemoryLimitHandle().Run()
+
+	tracker1.Consume(100 << 20) // 100 MB
+	tracker2.Consume(200 << 20) // 200 MB
+	tracker3.Consume(300 << 20) // 300 MB
+
+	test := make([]int, 128<<20) // Keep 1GB HeapInUse
+	time.Sleep(500 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (wshwsh12) [reply executor/executor_test.go:6095]: The check goroutine checks the memory usage every 100ms. The Sleep() make sure that Top1Tracker can be Canceled.

File: executor/executor_test.go [hunk 12]
```diff
@@ -6063,3 +6063,66 @@ func TestTableLockPrivilege(t *testing.T) {
...
+		PS: []*util.ProcessInfo{tk1.Session().ShowProcess(), tk2.Session().ShowProcess(), tk3.Session().ShowProcess()},
+	}
+	dom.ServerMemoryLimitHandle().SetSessionManager(sm)
+	go dom.ServerMemoryLimitHandle().Run()
+
+	tracker1.Consume(100 << 20) // 100 MB
+	tracker2.Consume(200 << 20) // 200 MB
+	tracker3.Consume(300 << 20) // 300 MB
+
+	test := make([]int, 128<<20) // Keep 1GB HeapInUse
```
[the remark this answers is not in the document]
Comment (wshwsh12) [reply executor/executor_test.go:6094]: This only simulates the situation of `heapinuse` growth. 
In normally, TiDB will use the memory immediately after allocation.

File: executor/executor_test.go [hunk 13]
```diff
@@ -6063,3 +6063,66 @@ func TestTableLockPrivilege(t *testing.T) {
...
+	require.True(t, tracker3.NeedKill.Load())
+	require.Equal(t, memory.MemUsageTop1Tracker.Load(), tracker3)
+	util.WithRecovery( // Next Consume() will panic and cancel the SQL
+		func() {
+			tracker3.Consume(1)
+		}, func(r interface{}) {
+			require.True(t, strings.Contains(r.(string), "Out Of Memory Quota!"))
+		})
+	tracker2.Consume(300 << 20) // Sum 500MB, Not Panic, Waiting t3 cancel finish.
+	time.Sleep(500 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (wshwsh12) [reply executor/executor_test.go:6109]: Wait for cancel

File: sessionctx/variable/sysvar.go [hunk 14]
```diff
@@ -716,6 +716,54 @@ var defaultSysVars = []*SysVar{
...
+		Validation: func(s *SessionVars, normalizedValue string, originalValue string, scope ScopeFlag) (string, error) {
+			intVal, err := strconv.ParseUint(normalizedValue, 10, 64)
+			if err != nil {
+				return "", err
+			}
+			if intVal > 0 && intVal < (512<<20) { // 512 MB
+				s.StmtCtx.AppendWarning(ErrTruncatedWrongValue.GenWithStackByArgs(TiDBServerMemoryLimit, originalValue))
+				intVal = 512 << 20
+			}
+			return strconv.FormatUint(intVal, 10), nil
```
[the remark this answers is not in the document]
Comment (wshwsh12) [reply sessionctx/variable/sysvar.go:732]: 0 means disable this feature.

File: executor/executor_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/executor_test.go:6095]: It's better to add some comment for it.
Sleep slows down the test, and 0.5s is long enough for UT

File: sessionctx/variable/sysvar.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply sessionctx/variable/sysvar.go:732]: OK... I see. This is kind of confusing
0 means disable, while when the value is smaller than 512MB, it's adjusted to 512MB (actual min value)

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**